cnosdb/src/main/java/cn/edu/tsinghua/iot/benchmark/cnosdb/CnosDB.java

```diff
@@ -0,0 +1,555 @@
+import cn.edu.tsinghua.iot.benchmark.schema.schemaImpl.DeviceSchema;
+import cn.edu.tsinghua.iot.benchmark.tsdb.DBConfig;
+import cn.edu.tsinghua.iot.benchmark.tsdb.IDatabase;
+import cn.edu.tsinghua.iot.benchmark.tsdb.TsdbException;
+import cn.edu.tsinghua.iot.benchmark.utils.TimeUtils;
+import cn.edu.tsinghua.iot.benchmark.workload.query.impl.*;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import okhttp3.OkHttpClient.Builder;
+import okhttp3.Response;
+import org.influxdb.dto.BatchPoints;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+public class CnosDB implements IDatabase {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CnosDB.class);
+  private static Config config = ConfigDescriptor.getInstance().getConfig();
+
+  private final String cnosUrl;
+  private final String cnosDbName;
+
+  private org.influxdb.InfluxDB influxDbInstance;
+  private CnosConnection cnosConnection;
+  private static final long TIMESTAMP_TO_NANO = getToNanoConst(config.getTIMESTAMP_PRECISION());
+
+  /** constructor. */
+  public CnosDB(DBConfig dbConfig) {
+    cnosUrl = "http://" + dbConfig.getHOST().get(0) + ":" + dbConfig.getPORT().get(0);
+    cnosDbName = dbConfig.getDB_NAME();
+  }
+
+  @Override
+  public void init() throws TsdbException {
+    try {
+      Builder client =
+          new Builder()
+              .connectTimeout(5, TimeUnit.MINUTES)
+              .readTimeout(5, TimeUnit.MINUTES)
+              .writeTimeout(5, TimeUnit.MINUTES)
+              .retryOnConnectionFailure(true);
+      influxDbInstance = org.influxdb.InfluxDBFactory.connect(cnosUrl, client);
+      cnosConnection = new CnosConnection(cnosUrl, cnosDbName);
+    } catch (Exception e) {
+      LOGGER.error("Initialize CnosDB failed because ", e);
+      throw new TsdbException(e);
+    }
+  }
+
+  @Override
+  public void cleanup() throws TsdbException {
+    try {
+      Response response = cnosConnection.execute("DROP DATABASE IF EXISTS " + cnosDbName);
+      response.close();
+    } catch (Exception e) {
+      LOGGER.error("Cleanup CnosDB failed because ", e);
+      throw new TsdbException(e);
+    }
+  }
+
+  @Override
+  public void close() {
+    if (influxDbInstance != null) {
+      influxDbInstance.close();
+    }
+  }
+
+  @Override
+  public Double registerSchema(List<DeviceSchema> schemaList) throws TsdbException {
+    long start;
+    long end;
+    try {
+      start = System.nanoTime();
+      Response response =
+          cnosConnection.execute(
+              String.format("create database if not exists %s with shard 32", cnosDbName));
```

Comment:
Maybe it would be better to separate out the number of shards and make it a configurable parameter ~ but I don't have a problem with this implementation either